File: boj/1753.cpp
#include <stdio.h>
#include <queue>
#include <vector>
using namespace std;
const long INF= 987654321;
typedef pair<long,long> pi;

int main(){
	queue <long> q;	
	vector<pi> G[20001];
	
	long long V,E,start;
	scanf("%lld %lld %lld",&V,&E,&start);
	
	for (int i=0; i<E;i++){
		int u,v,w;
		scanf("%d %d %d",&u,&v,&w);
		G[u].push_back(pi(v,w));

	}
	vector<long> distance(V+1,INF);
	distance[start]=0;
	q.push(start);
	
	while(!q.empty()){
		long cur=q.front();
		q.pop();
		for(int i=0;i<G[cur].size();i++){
			long n=G[cur][i].first;
			long w=G[cur][i].second;
			if( distance[n]>distance[cur]+w){
				distance[n]=distance[cur]+w;
				q.push(n);
			}
		}
		
	}
	for(int i=1;i<V+1;i++){
		if( distance[i]==INF)
			printf("INF\n");
		else 
			printf("%ld\n",distance[i]);
	}
	return 0;
}

File: boj/1753.py
import sys
import Queue
q=Queue.deque()

v,e=map(int,sys.stdin.readline().split())
INF=987654321
start=int(sys.stdin.readline())
g=[Queue.deque() for i in range(v+1)]
distance=[INF]*(v+1)
   
for i in range(e):
    u,v,w=map(int,sys.stdin.readline().split())
    g[u].appendleft((v,w))
   

distance[start]=0
q.appendleft(start)

while(len(q)>0):
    cur=q.pop()

    for n,w in g[cur]:
        if distance[n]>distance[cur]+w:
            distance[n]=distance[cur]+w
            q.appendleft(n)
   
   
for i in distance[1:]:
    if i==INF:
        print("INF")
    else:print(i)